NetApp replica promote: guard source being down

Carthaca opened the pull request:
Constructing a vserver client to the old active replica
may fail in a disaster scenario. Promoting the non-active
replica must work nevertheless.

Change-Id: Ib1a36f45ba47e378e13acd24ea16634ed96e15ca

Carthaca committed:
NetApp replica promote: guard source being down

Constructing a vserver client to the old active replica
may fail in a disaster scenario. Promoting the non-active
replica must work nevertheless.

Change-Id: Ib1a36f45ba47e378e13acd24ea16634ed96e15ca

chuan137 commented:
I just noticed that the code changes we did in the fire drill was from changing the snapmirror source to the new active replica. I am wondering is such a disaster situation, there is better way to handle it than error it out. Can we forcefully cut the old snapmirror relationships and elect the new replica? It might be helpful when there are more than two replicas

chuan137 commented on manila/share/drivers/netapp/dataontap/cluster_mode/lib_base.py:
```diff
@@ -3334,24 +3351,17 @@ def _safe_change_replica_source(self, dm_session, replica,
-            LOG.exception(msg, replica['id'])
-            return replica
-        except netapp_api.NaApiError:
+        # Catch everything! The original source may be down,
```
I am wondering is such a disaster situation, there is better way to handle it than error it out. Can we forcefully cut the old snapmirror relationships and elect the new replica? It might be helpful when there are more than two replicas

Carthaca committed:
snapmirror break: wait for abort to be done on replica promote

The 'aborting' takes a while until snapmirror break can continue,
give it 2 minutes at most.

Change-Id: I0b411a00ab1fd72756e60e64cd9a71482a7998be

Carthaca commented:
rebased as #300
